pkg/proc/native/threads_linux_arm.go

```diff
@@ -0,0 +1,237 @@
+package native
+
+import (
+	"debug/elf"
+	"encoding/binary"
+	"fmt"
+	"golang.org/x/arch/arm/armasm"
+	"math/bits"
+	"syscall"
+	"unsafe"
+
+	sys "golang.org/x/sys/unix"
+
+	"github.com/go-delve/delve/pkg/proc"
+	"github.com/go-delve/delve/pkg/proc/linutil"
+)
+
+func (t *nativeThread) fpRegisters() ([]proc.Register, []byte, error) {
+	var err error
+	var arm_fpregs linutil.ARM64PtraceFpRegs
+	t.dbp.execPtraceFunc(func() { arm_fpregs.Vregs, err = ptraceGetFpRegset(t.ID) })
+	fpregs := arm_fpregs.Decode()
+	if err != nil {
+		err = fmt.Errorf("could not get floating point registers: %v", err.Error())
+	}
+	return fpregs, arm_fpregs.Vregs, err
+}
+
+func (t *nativeThread) restoreRegisters(savedRegs proc.Registers) error {
+	sr := savedRegs.(*linutil.ARMRegisters)
+
+	var restoreRegistersErr error
+	t.dbp.execPtraceFunc(func() {
+		restoreRegistersErr = ptraceSetGRegs(t.ID, sr.Regs)
+		if restoreRegistersErr != syscall.Errno(0) {
+			return
+		}
+		if sr.Fpregset != nil {
+			iov := sys.Iovec{Base: &sr.Fpregset[0], Len: uint32(len(sr.Fpregset))}
+			_, _, restoreRegistersErr = syscall.Syscall6(syscall.SYS_PTRACE, sys.PTRACE_SETREGSET, uintptr(t.ID), uintptr(elf.NT_FPREGSET), uintptr(unsafe.Pointer(&iov)), 0, 0)
+		}
+	})
+	if restoreRegistersErr == syscall.Errno(0) {
+		restoreRegistersErr = nil
+	}
+	return restoreRegistersErr
+}
+
+// resolvePCForArm is used to resolve all next PC for current instruction.
+func (t *nativeThread) resolvePC(regs proc.Registers) ([]uint64, error) {
+	// Use ptrace to get better performance.
+	nextInstrLen := t.BinInfo().Arch.MaxInstructionLength()
+	nextInstrBytes := make([]byte, nextInstrLen)
+	var err error
+	t.dbp.execPtraceFunc(func() {
```

Comment:
Why not just use t.ReadMemory (here and in the other places in this function where PtracePeekData is called)?

Reply:
As this is inside low level emulation of ptrace singlestep, I want to make sure it first enter the context of ptrace channel, then do the peek and poke work, so I skip the upper Read/WriteMemory function, and use ptrace function directly.